main/src/ca/uwaterloo/flix/language/phase/unification/ContextReduction.scala

```diff
@@ -0,0 +1,115 @@
+
+import scala.annotation.tailrec
+
+object ContextReduction {
+
+  /**
+    * Returns `true` iff type constraints `tconstrs0` entail type constraint `tconstr`, under class environment `instances`.
+    * That is, `tconstr` is true if all of `tconstrs0` are true.
+    */
+  // MATT THIH says that toncstrs0 should always be in HNF so checking for byInst is a waste.
+  def entail(instances: MultiMap[Symbol.ClassSym, ResolvedAst.Instance], tconstrs0: List[TypedAst.TypeConstraint], tconstr: TypedAst.TypeConstraint)(implicit flix: Flix): Boolean = {
+    // tconstrs0 entail tconstr if tconstr is a superclass of any member or tconstrs0
+    // or if there is an instance matching tconstr and all of the instance's constraints are entailed by tconstrs0
+    tconstrs0.exists(bySuper(instances, _).contains(tconstr)) || {
+      byInst(instances, tconstr) match {
+        case Result.Ok(tconstrs) => tconstrs.forall(entail(instances, tconstrs0, _))
+        case Result.Err(_) => false
+      }
+    }
+  }
+
+  /**
+    * Removes the type constraints which are entailed by the others in the list.
+    */
+  private def simplify(instances: MultiMap[Symbol.ClassSym, ResolvedAst.Instance], tconstrs0: List[TypedAst.TypeConstraint])(implicit flix: Flix): List[TypedAst.TypeConstraint] = {
+
+    @tailrec
+    def loop(acc: List[TypedAst.TypeConstraint], tconstrs0: List[TypedAst.TypeConstraint]): List[TypedAst.TypeConstraint] = tconstrs0 match {
+      case Nil => acc
+      case head :: tail if entail(instances, acc ++ tail, head) => loop(acc, tail)
+      case head :: tail => loop(head :: acc, tail)
+    }
+
+    loop(Nil, tconstrs0)
+  }
+
+  /**
+    * Normalizes a list of type constraints, converting to head-normal form and removing semantic duplicates.
+    */
+  private def reduce(instances: MultiMap[Symbol.ClassSym, ResolvedAst.Instance], tconstrs0: List[TypedAst.TypeConstraint])(implicit flix: Flix): Result[List[TypedAst.TypeConstraint], UnificationError] = {
+    for {
+      tconstrs <- Result.sequence(tconstrs0.map(toHeadNormalForm(instances, _)))
+    } yield simplify(instances, tconstrs.flatten)
+  }
+
+  /**
+    * Converts the type constraint to head-normal form, i.e. `a[X1, Xn]`, where `a` is a variable and `n >= 0`.
+    */
+  private def toHeadNormalForm(instances: MultiMap[Symbol.ClassSym, ResolvedAst.Instance], tconstr: TypedAst.TypeConstraint)(implicit flix: Flix): Result[List[TypedAst.TypeConstraint], UnificationError] = {
+    if (isHeadNormalForm(tconstr.arg)) {
+      List(tconstr).toOk
+    } else {
+      byInst(instances, tconstr)
+    }
+  }
+
+  /**
+    * Returns the list of constraints that hold if the given constraint `tconstr` holds, using the constraints on available instances.
+    */
+  private def byInst(instances0: MultiMap[Symbol.ClassSym, ResolvedAst.Instance], tconstr: TypedAst.TypeConstraint)(implicit flix: Flix): Result[List[TypedAst.TypeConstraint], UnificationError] = {
```

Comment:
I would swap the arguments, because in general our type signatures look like:

"the thing we are most concerned with/working on", ... "lots of auxilliary information, e.g. root, instances,etc."